Add more tests for receiver check in IntegerIndexedElementSet

anba opened the pull request:
Additional tests to cover some spec-compliance issues I've encountered while implementing tc39/ecma262#1556 in SpiderMonkey.

Ms2ger commented:
Did you mean to push all three commits to this PR? I'm happy to approve the first one, at least, but would need to spend some more time reviewing the others.

anba commented:
Yes, that was intentional, because at least for SpiderMonkey there are [shared code paths](https://searchfox.org/mozilla-central/rev/490a1df802d8872f996f8ef4093d54e3c854c8f9/js/src/vm/NativeObject.h#1838-1850) for normal property operations and environment operations, so when I've updated the property operations for out-of-bounds typed array indices, I also had to worry about environment operations. Similar story about the new test for `TypedArray.prototype.set`: I've also updated the [TypedArraySetElement](https://tc39.es/ecma262/#sec-typedarraysetelement) implementation in SM, which led to checking which spec operations actually call `TypedArraySetElement` and writing some small tests to check implementations agree on the spec semantics. This uncovered a spec violation in JSC (72933dfc0be2394956cf29689336dd7678d089c0).
